### ISSUE TITLE
report_similar_impl_candidates seems to not put things in a canonical order

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust#53302]
[I hit](https://github.com/rust-lang/rust/pull/52602) the following UI test failure in travis today: 
```diff
 	   |         ^^^^^^^^ the trait `std::convert::From<&str>` is not implemented for `i32`
 	   |
 	   = help: the following implementations were found:
+	             <i32 as std::convert::From<u8>>
 	             <i32 as std::convert::From<i8>>
-	             <i32 as std::convert::From<bool>>
 	             <i32 as std::convert::From<u16>>
 	             <i32 as std::convert::From<i16>>
-	             <i32 as std::convert::From<u8>>
+	             <i32 as std::convert::From<bool>>
 	   = note: required by `std::convert::From::from`
```
https://travis-ci.org/rust-lang/rust/jobs/415272826#L1942-L1958

Despite being fully-up-to-date against master and that passing locally.

Perhaps the order is deterministic normally, but different for me because I built `--incremental`?

I figured it was worth filing, at least, as I can't see any obvious sort in the relevant code:

https://github.com/rust-lang/rust/blob/5a0d2961ce88f7db90b13771d1e8fc3b50ded7b1/src/librustc/traits/error_reporting.rs#L407-L470

**Comment by estebank** [comment]
In the code linked above, modify `impl_candidates` to be `mut` and sort it first thing in the method `report_similar_impl_candidates`. You might have to sort the normalized values though, as `TraitRef` might 1) not be sortable and 2) might sort differently in different targets. Because of this it would make sense to sort on the display value for the fields. This will be less efficient, but we're already reporting an error at this point, so it's ok.

**Comment by BurntPizza** [comment]
I can take this.

**Comment by BurntPizza** [comment]
This should be fixed now by #53196.
